changeset.go

```diff
@@ -109,63 +109,94 @@ func (cs *Changeset) Upsert(ctx context.Context, changesetNamespace, changesetNa
 			}
 			return trace.Wrap(err)
 		}
-		err = cs.upsertResource(ctx, changesetNamespace, changesetName, raw.Raw)
+		err = cs.upsertResource(ctx, changesetNamespace, changesetName, id, raw.Raw)
 		if err != nil {
 			return trace.Wrap(err)
 		}
 	}
 }
 
-func (cs *Changeset) upsertResource(ctx context.Context, changesetNamespace, changesetName string, data []byte) error {
+func (cs *Changeset) upsertResource(ctx context.Context, changesetNamespace, changesetName, id string, data []byte) error {
 	tr, err := cs.createOrRead(changesetNamespace, changesetName, ChangesetSpec{Status: ChangesetStatusInProgress})
 	if err != nil {
 		return trace.Wrap(err)
 	}
 	if tr.Spec.Status != ChangesetStatusInProgress {
 		return trace.CompareFailed("cannot update changeset - expected status %q, got %q", ChangesetStatusInProgress, tr.Spec.Status)
 	}
-	var kind metav1.TypeMeta
-	err = yaml.NewYAMLOrJSONDecoder(bytes.NewReader(data), DefaultBufferSize).Decode(&kind)
+	headers, err := ParseResourceHeader(bytes.NewReader(data))
 	if err != nil {
 		return trace.Wrap(err)
 	}
-	switch kind.Kind {
+
+	comp, err := cs.isOpCompleted(ctx, tr, id)
+	if comp {
+		log := log.WithFields(log.Fields{
+			"cs": tr.String(),
+		})
+		log.Infof("%s '%s' is already updated.", headers.TypeMeta.Kind, headers.ObjectMeta.Name)
+		return nil
```

Comment:
Should we allow `--force` flag to overrule the "already completed" status and upsert anyway? Otherwise there seems to be no way to force-retry. Same for delete.

Reply:
As I mentioned in one of my comments you can't just retry operation without losing original resource.